components/supersaas/sources/changed-appointments.js

```diff
@@ -0,0 +1,39 @@
+const supersaas = require('../supersaas.app.js');
+
+module.exports = {
+  key: 'supersaas-changed-appointments',
+  name: 'New or changed appointments',
+  description: `This event source tracks new and changed appointment events from the selected schedules and emits them.`,
+  version: '0.0.1',
+
+  props: {
+    supersaas,
+    schedules: { propDefinition: [supersaas, 'schedules'] },
+    db: "$.service.db",
+    http: '$.interface.http',
+  },
+
+  hooks: {
+    async activate() {
+      const { http, schedules } = this;
+
+      this.db.set('activeHooks', await this.supersaas.createHooks(schedules.map(x => ({
+        event: 'C', // change_appointment
+        parent_id: Number(x),
+        target_url: http.endpoint,
+      }))));
+    },
+
+    async deactivate() {
+      await this.supersaas.destroyHooks(this.db.get('activeHooks') || []);
+      this.db.set('activeHooks', []);
+    },
+  },
+
+  async run(ev) {
+    // TODO: Authenticate requests? (see GitHub's HMAC signature approach)
+
+    console.log('Emitting:', ev.body);
+    this.$emit(ev.body);
```

Comment:
Did you happen to see the metadata you can pass to the emitted events? https://github.com/PipedreamHQ/pipedream/blob/master/COMPONENT-API.md#emit

At a minimum `summary` is normally nice to include, since that's what appears in the UI for the emitted event for the user and can make it easier to identify the event. In this case if you have access to the appointment name, that might be nice to include as the summary.

`id` is nice to include if you want to take advantage of the [built-in deduper](https://github.com/PipedreamHQ/pipedream/blob/master/COMPONENT-API.md#dedupe-strategies). If you include a unique event ID (fixed for a particular webhook request) in the payload, for example, you can pass that as the `id`, then define a `dedupe: unique` strategy in the component, and we'll make sure duplicate events don't get emitted. Optional but a nice feature.

Reply:
Resource names are sometimes available but mostly not. Our webhooks seem to be sending slightly inconsistent event structures in this regard, but we don't want to make changes to our codebase to support this for now because we don't know yet how worthwhile is this whole effort (only very few users want it in principle) and definitely don't want to accidentally cause trouble to many other customers.

But there is enough information in standard format to write a summary, yes, so I'll do that.

This will probably mean moving some of the code I wrote for the logging action step into the event source code for `credit-balance-changes`.

Reply:
👍